infra/Pos/Communication/Relay/Class.hs

```diff
@@ -1,60 +1,82 @@
                                                  ReqMsg (..))
 
--- | Typeclass for general Inv/Req/Dat framework. It describes monads,
--- that store data described by tag, where "key" stands for node
--- identifier.
-class ( Buildable tag
-      , Buildable contents
+-- | Data for general Inv/Req/Dat framework
+
+data Relay m where
+  InvReqData ::
+      ( Buildable contents
       , Buildable key
-      , Typeable tag
       , Typeable contents
       , Typeable key
-      , Message (ReqMsg key tag)
-      , Message (MempoolMsg tag)
-      , Message (InvOrData tag key contents)
+      , Eq key
+      , Bi (ReqMsg key)
+      , Bi (InvOrData key contents)
+      , Message (ReqMsg key)
+      , Message (InvOrData key contents)
+      , MessageLimited (DataMsg contents)
+      ) => MempoolParams m -> InvReqDataParams key contents m -> Relay m
+  Data ::
+      ( Buildable contents
+      , Typeable contents
+      , Bi (DataMsg contents)
+      , Message (DataMsg contents)
       , MessageLimited (DataMsg contents)
-      ) => Relay m tag key contents
-      | tag -> contents, contents -> tag, contents -> key, tag -> key where
-    -- | Converts data to tag. Tag returned in monad `m`
-    -- for only type matching reason (multiparam type classes are tricky)
-    contentsToTag :: contents -> m tag
+      ) => DataParams contents m -> Relay m
+
+data MempoolParams m where
+    NoMempool :: MempoolParams m
+    -- `tag` is used only as type param, no actual param used
+    KeyMempool ::
+      ( Message (MempoolMsg tag)
+      , Message (InvMsg key)
+      , Bi (MempoolMsg tag)
+      , Bi (InvMsg key)
+      , Buildable key
+      , Typeable tag
+      , Typeable key
+      ) => Proxy tag -> m [key] -> MempoolParams m
 
-    -- | Same for key. Sometime contents has key inside already, so
-    -- it's redundant to double-pass it everywhere.
-    contentsToKey :: contents -> m key
+data InvReqDataParams key contents m =
+    InvReqDataParams
+        {
+        -- | Same for key. Sometime contents has key inside already, so
+        -- it's redundant to double-pass it everywhere.
+          contentsToKey :: contents -> m key
 
-    verifyInvTag :: tag -> m VerificationRes
-    verifyReqTag :: tag -> m VerificationRes
-    verifyMempoolTag :: tag -> m VerificationRes
-    verifyDataContents :: contents -> m VerificationRes
+        -- | Handle inv msg and return whether it's useful or not
+        , handleInv     :: key -> m Bool
 
-    -- | Handle inv msg and return whether it's useful or not
-    handleInv :: tag -> key -> m Bool
+        -- | Handle req msg and return (Just data) in case requested data can be provided
+        , handleReq     :: key -> m (Maybe contents)
 
-    -- | Handle req msg and return (Just data) in case requested data can be provided
-    handleReq :: tag -> key -> m (Maybe contents)
+        -- | Handle data msg and return True if message is to be propagated
+        , handleData    :: contents -> m Bool
```

Comment:
It would be good to return something more than `Bool`, some reason why we don't like given data. Currently we can just log it (it's very valuable information and it's better to log it in one place, not four). Later we will probably add `reject` message and we can send this information as a payload.